text/0000-unix-socket.md

```diff
@@ -0,0 +1,491 @@
+}
+
+impl IntoRawFd for UnixDatagram {
+    ...
+}
+```
+
+Differences from `UdpSocket`:
+* `bind` takes an `AsRef<Path>` rather than a `ToSocketAddrs`.
+* The `unbound` method creates an unbound socket, as a Unix socket does not need
+  to be bound to send messages.
+* The `pair` method creates a pair of connected, unnamed sockets, as this is
+  commonly used for IPC.
+* The `SocketAddr` returned by the `local_addr` and `peer_addr` methods is
+  different.
+* The `connect`, `send`, `recv`, `set_nonblocking`, and `take_error` methods are
+  not currently present on `UdpSocket` but are provided in the `net2` crate and
+  are being proposed for addition to the standard library in a separate RFC.
+
+## Functionality not present
+
+Some functionality is notably absent from this proposal:
+
+* No support for `SOCK_SEQPACKET` sockets is proposed, as it has not yet been
+  implemented. Since it is connection oriented, there will be a socket type
+  `UnixSeqPacket` and a listener type `UnixSeqListener`. The naming of the
+  listener is a bit unfortunate, but use `SOCK_SEQPACKET` is rare compared to
+  `SOCK_STREAM` so naming priority can go to that version.
+* Unix sockets support file descriptor and credential transfer, but these will
+  not initially be supported as the `sendmsg`/`recvmsg` interface is complex
+  and bindings will need some time to prototype.
+
+These features can bake in the `rust-lang-nursery/unix-socket` as they're
+developed.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+While there is precedent for platform specific components in the standard
+library, this will be the by far the largest platform specific addition.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Unix socket support could be left out of tree.
+
+The naming convention of `UnixStream` and `UnixDatagram` doesn't perfectly
+mirror `TcpStream` and `UdpSocket`, but `UnixStream` and `UnixSocket` seems way
+too confusing.
+
+Constructors for the various socket types take an `AsRef<Path>`, which makes
+construction of sockets associated with Linux abstract namespaces somewhat
+nonobvious, as the leading null byte has to be explicitly added. However, it is
+still possible, either via `&str` for UTF8 names or via `&OsStr` and
+`std::os::unix::ffi::OsStrExt` for arbitrary names. Use of the abstract
+namespace appears to be very obscure, so it seems best to optimize for
+ergonomics of normal pathname addresses. We can add extension traits providing
+methods taking `&[u8]` in the future if deemed necessary.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Is `std::os::unix::net` the right name for this module? It's not strictly
+"networking" as all communication is local to one machine. `std::os::unix::unix`
+is more accurate but weirdly repetitive and the extension trait module
+`std::os::linux::unix` is even weirder. `std::os::unix::socket` is an option,
+but seems like too general of a name for specifically `AF_UNIX` sockets as
+opposed to *all* sockets.
```

Comment:
Perhaps another unresolved question for now could be whether Windows named pipes should be added in parallel? Or something along the lines of if we should hold off on stabilization until there's a suitable Windows alternative as well.